src/vocgui/views/training_set_by_id_viewset.py

```diff
@@ -0,0 +1,47 @@
+from rest_framework import viewsets
+from django.db.models import Count, Q
+from vocgui.models import TrainingSet
+from vocgui.serializers import TrainingSetSerializer
+from .utils import check_group_object_permissions
+from django.http import Http404
+
+
+class TrainingSetByIdViewSet(viewsets.ReadOnlyModelViewSet):
+    """
+    Defines a view set for a TrainingSet module of a given id.
+    Inherits from `viewsets.ReadOnlyModelViewSet` and defines queryset
+    and serializers.
+    """
+
+    serializer_class = TrainingSetSerializer
+
+    def get_queryset(self):
+        """
+        Defining custom queryset
+
+        :param self: A handle to the :class:`TrainingSetViewSet`
+        :type self: class
+
+        :return: (filtered) queryset
+        :rtype: QuerySet
+        """
+        if getattr(self, "swagger_fake_view", False):
+            return TrainingSet.objects.none()
+        user = self.request.user
+        queryset = TrainingSet.objects.filter(
+            id=self.kwargs["pk"], released=True
+        ).annotate(
+            total_documents=Count(
+                "documents",
+                filter=Q(documents__document_image__confirmed=True),
+                distinct=True,
+            )
+        )
+        try:
+            group_id = getattr(queryset[0], "created_by_id")
+            is_admin = getattr(queryset[0], "creator_is_admin")
+        except:
+            raise Http404
+        if group_id and not is_admin:
+            check_group_object_permissions(self.request, group_id)
```

Comment:
```suggestion
        key = get_key(self.request)
        if key:
            try:
                # If an API key was provided, only return training sets that were created by the corresponding group
                queryset = queryset.filter(
                    created_by=GroupAPIKey.objects.get_from_key(key).organization
                )
            except GroupAPIKey.DoesNotExist:
                raise PermissionDenied()
        else:
            # If no API key was provided, only return training sets created by admins
            queryset = queryset.filter(creator_is_admin=True)
```
I would suggest not to check the permissions on the object level, because it is hard to combine it with a working list view. Instead, we could filter the queryset to return either all training sets of a specific group (if an API key is provided) or all training sets created by admins (if no key is given). Appending the id will then automatically retrieve the object if it is included in the queryset, which means that permissions are automatically checked for the detail view as well.

The only difference in the behaviour would be that if a private training set is requested without an API key, a 404 is returned instead of 403. If that's a problem we'd have to overwrite the `get_object()` function of the `ReadOnlyModelViewSet`.

If you want to, we could also return the public training sets together with the private ones when an API key given:
```
# If an API key was provided, return training sets that were created either by the corresponding group or by admins
queryset = queryset.filter(
    Q(created_by=GroupAPIKey.objects.get_from_key(key).organization)
    | Q(creator_is_admin=True)
)
```

Reply:
Yes, currently admin training sets return 404 when an API key is given, but requesting both private and admin trainingsets with an API key is very easy by just adding the admin trainingsets to the queryset in this case (see my comment above), so if you would prefer that behavior, you could just change that line as well.

But yes, I agree with putting a bit of thought in the 403/404 question for the next API version!